### ISSUE TITLE
Ability to run Action locally (offline) on a manifest during development

### ISSUE THREAD

**Issue by nazarewk** [argoproj/argo-cd#10703]
# Summary

I think it would be great to be able to test actions locally:
1. without uploading it to server (`argocd-cm`)
2. on a specific resource manifest (stdin/local file)

# Motivation

Currently it's not possible to test actions in a restricted ArgoCD interpreter. Using separate Lua 5.1 is far from ideal.

# Proposal

Action could be runnable using the same command [`argocd app actions run`](https://argo-cd.readthedocs.io/en/stable/user-guide/commands/argocd_app_actions_run/) extended by new parameters to modify it's behaviour.

**Comment by jessesuen** [comment]
I think it's an interesting idea. This would be a CLI only feature and would leverage our existing resource edit feature. It would only run the lua interpreter to modify an object. 

From the perspective of the API server, it will not know that an action was invoked, but only see it as a resource edit.

One potential issue is that it's possible this could impact the `argocd` CLI size since it will pull in the lua interpreter and dependencies. Also, I expect there to be some non-trivial amount of work refactoring the resource customization/action package to make this importable/usable from the CLI.

**Comment by nazarewk** [comment]
I have a poor man's substitute using `lyaml` and `std.io` libraries:

<details>
<summary> run.lua </summary>


```lua
#!/usr/bin/env lua5.1
-- usage: lua5.1 ./run.lua [function_code_file=action.lua] [k8s_manifest=obj.yaml]

-- a poor substitute for https://github.com/argoproj/argo-cd/issues/10703 :
-- 1) using 2 libraries it loads an object from obj.yaml
-- 2) loads a function body from a lua file (eg: copy-pasted `discovery.lua` or `action.lua` from `argocd-cm.yaml`)
-- 3) dumps result formatted as YAML to stdout

-- caveats:
-- 1) it does not restrict function environment like Argo CD does, setfenv() was implemented in 5.2
--    see https://github.com/argoproj/argo-cd/blob/b431d253e5f8655ec6285cabed008ce45e34d4f9/util/lua/lua.go#L58-L92
-- 2) see LuaRocks for installing packages https://luarocks.org/
-- 3) must be run by Lua 5.1 interpreter due to Argo CD using https://github.com/yuin/gopher-lua

assert(_VERSION == "Lua 5.1", "You must use Lua in version 5.1 due to Argo CD using https://github.com/yuin/gopher-lua")

local lyaml = require("lyaml")
local io = require("std.io")
local code, manifest = ...

code = code or "action.lua"
manifest = manifest or "obj.yaml"

action_body = loadfile(code)

obj = lyaml.load(io.slurp(manifest))
result = action_body(obj)
print(_VERSION)
print(lyaml.dump({ result }))

```

</details>

**Comment by nazarewk** [comment]
> One potential issue is that it's possible this could impact the `argocd` CLI size since it will pull in the lua interpreter and dependencies. Also, I expect there to be some non-trivial amount of work refactoring the resource customization/action package to make this importable/usable from the CLI.

you are right, could be measured and/or worked around by introducing a separate binary?

**Comment by nazarewk** [comment]
> @nazarewk I'm not super familiar with actions. Would the CLI _need_ to write the changes to the API server? Or could it just spit out the modified resource as YAML and let the user diff it?

not at all, just spit out results to either stdout or file
